File: concurrency-and-node/haskell-async/src/RemoteMain.hs
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE OverloadedStrings #-}

import Control.Concurrent.Async
import Control.Lens
import Data.Aeson.Lens
import Data.Time
import Network.Wreq

main :: IO ()
main = do
  testWithoutAsync
  testWithAsync

testWithoutAsync :: IO ()
testWithoutAsync = do
  !t1 <- getCurrentTime
  (Just x) <- (^? responseBody . key "fibResult" . _Integral) <$> get "http://localhost:8080/api/fibs/43"
  (Just y) <- (^? responseBody . key "fibResult" . _Integral) <$> get "http://localhost:8080/api/fibs/44"
  let sum = (x :: Int) + (y :: Int)
  !t2 <- getCurrentTime
  putStrLn $ "Without Async"
  putStrLn $ "Sum: " ++ show sum
  putStrLn $ "Execution Time: " ++ show (diffUTCTime t2 t1)
{-# NOINLINE testWithoutAsync #-}

testWithAsync :: IO ()
testWithAsync = do
  !t1 <- getCurrentTime
  (Just x, Just y) <- concurrently
            ((^? responseBody . key "fibResult" . _Integral) <$> get "http://localhost:8080/api/fibs/43")
            ((^? responseBody . key "fibResult" . _Integral) <$> get "http://localhost:8080/api/fibs/44")
  let sum = (x :: Int) + (y :: Int)
  !t2 <- getCurrentTime
  putStrLn $ "With Async"
  putStrLn $ "Sum: " ++ show sum
  putStrLn $ "Execution Time: " ++ show (diffUTCTime t2 t1)
{-# NOINLINE testWithAsync #-}

File: concurrency-and-node/node-async/src/remote-main.js
async = require('promise-async');
request = require('request-promise');

function main() {
    testWithoutAsync()
        .then(() => {
            return testWithAsync();
        });
}

function testWithoutAsync() {
    var x;
    var y;
    var t1 = new Date().getTime();

    return request('http://localhost:8080/api/fibs/43')
        .then((xStr) => {
            x = JSON.parse(xStr).fibResult;
            return request('http://localhost:8080/api/fibs/44');
        })
        .then((yStr) => {
            y = JSON.parse(yStr).fibResult;

            var sum = x + y;
            var t2 = new Date().getTime();

            console.log('Without Async');
            console.log('Sum: ' + sum);
            console.log('Execution Time: ' + ((t2 - t1) / 1000) + 's');
        })
}

function testWithAsync() {
    var x;
    var y;
    var t1 = new Date().getTime();

    return async.parallel([
        (callback) => {
            request('http://localhost:8080/api/fibs/43')
                .then((xStr) => {
                    x = JSON.parse(xStr).fibResult;
                    callback();
                });
        },
        (callback) => {
            request('http://localhost:8080/api/fibs/44')
                .then((yStr) => {
                    y = JSON.parse(yStr).fibResult;
                    callback();
                });
        }
    ]).then(() => {
        var sum = x + y;
        var t2 = new Date().getTime();

        console.log('With Async');
        console.log('Sum: ' + sum);
        console.log('Execution Time: ' + ((t2 - t1) / 1000) + 's');
    });
}

main();
